### ISSUE TITLE
v2.3 Compatability

### ISSUE THREAD

**Issue by venkatd** [neo4jrb/neo4j#1004]
Pulling this from #999 because it was unrelated to the actual bug in that one. Should be relevant for adding 2.3 compatibility though.

So I had run the tests on Neo4j 2.3 and looks like there were only 3 failing tests, and probably all related. My guess is that the type of error raised has just changed.

If you remove the `raise_error` expectation you'll see that it's still raising an error, just a different kind.

```
  1) Neo4j::ActiveNode::IdProperty id_property :my_id, on: :foobar throws exception if the same uuid is generated when saving node
     Failure/Error: expect { Clazz.create }.to raise_error(/Node \d+ already exists with label/)
       expected Exception with message matching /Node \d+ already exists with label/, got #<Neo4j::Server::Resource::ServerException: Expected response code 200 Error for request http://localhost:7474/db/data/transaction/3995, 404> with backtrace:
         # ./lib/neo4j/shared/callbacks.rb:39:in `create_model'
         # ./lib/neo4j/shared/persistence.rb:64:in `create_or_update'
         # ./lib/neo4j/shared/callbacks.rb:35:in `block in create_or_update'
         # ./lib/neo4j/shared/callbacks.rb:35:in `create_or_update'
         # ./lib/neo4j/active_node/persistence.rb:28:in `block in save'
         # ./lib/neo4j/active_node/persistence.rb:90:in `block in cascade_save'
         # ./lib/neo4j/active_node/persistence.rb:89:in `cascade_save'
         # ./lib/neo4j/active_node/persistence.rb:26:in `save'
         # ./lib/neo4j/shared/validations.rb:18:in `save'
         # ./lib/neo4j/active_node/persistence.rb:103:in `block in create'
         # ./lib/neo4j/active_node/persistence.rb:101:in `tap'
         # ./lib/neo4j/active_node/persistence.rb:101:in `create'
         # ./spec/e2e/id_property_spec.rb:189:in `block (4 levels) in <top (required)>'
         # ./spec/e2e/id_property_spec.rb:189:in `block (3 levels) in <top (required)>'
     # ./spec/e2e/id_property_spec.rb:189:in `block (3 levels) in <top (required)>'

  2) Neo4j::ActiveNode::IdProperty id_property :my_uuid, auto: :uuid throws exception if the same uuid is generated when saving node
     Failure/Error: expect { Clazz.create }.to raise_error(/Node \d+ already exists with label/)
       expected Exception with message matching /Node \d+ already exists with label/, got #<Neo4j::Server::Resource::ServerException: Expected response code 200 Error for request http://localhost:7474/db/data/transaction/4000, 404> with backtrace:
         # ./lib/neo4j/shared/callbacks.rb:39:in `create_model'
         # ./lib/neo4j/shared/persistence.rb:64:in `create_or_update'
         # ./lib/neo4j/shared/callbacks.rb:35:in `block in create_or_update'
         # ./lib/neo4j/shared/callbacks.rb:35:in `create_or_update'
         # ./lib/neo4j/active_node/persistence.rb:28:in `block in save'
         # ./lib/neo4j/active_node/persistence.rb:90:in `block in cascade_save'
         # ./lib/neo4j/active_node/persistence.rb:89:in `cascade_save'
         # ./lib/neo4j/active_node/persistence.rb:26:in `save'
         # ./lib/neo4j/shared/validations.rb:18:in `save'
         # ./lib/neo4j/active_node/persistence.rb:103:in `block in create'
         # ./lib/neo4j/active_node/persistence.rb:101:in `tap'
         # ./lib/neo4j/active_node/persistence.rb:101:in `create'
         # ./spec/e2e/id_property_spec.rb:305:in `block (4 levels) in <top (required)>'
         # ./spec/e2e/id_property_spec.rb:305:in `block (3 levels) in <top (required)>'
     # ./spec/e2e/id_property_spec.rb:305:in `block (3 levels) in <top (required)>'

  3) Neo4j::ActiveNode constraint constraint :name, type: :unique can not create two nodes with unique properties
     Failure/Error: expect { clazz_with_constraint.create(name: 'foobar') }.to raise_error StandardError, /already exists/
       expected StandardError with message matching /already exists/, got #<Neo4j::Server::Resource::ServerException: Expected response code 200 Error for request http://localhost:7474/db/data/transaction/4032, 404> with backtrace:
         # ./lib/neo4j/shared/callbacks.rb:39:in `create_model'
         # ./lib/neo4j/shared/persistence.rb:64:in `create_or_update'
         # ./lib/neo4j/shared/callbacks.rb:35:in `block in create_or_update'
         # ./lib/neo4j/shared/callbacks.rb:35:in `create_or_update'
         # ./lib/neo4j/active_node/persistence.rb:28:in `block in save'
         # ./lib/neo4j/active_node/persistence.rb:90:in `block in cascade_save'
         # ./lib/neo4j/active_node/persistence.rb:89:in `cascade_save'
         # ./lib/neo4j/active_node/persistence.rb:26:in `save'
         # ./lib/neo4j/shared/validations.rb:18:in `save'
         # ./lib/neo4j/active_node/persistence.rb:103:in `block in create'
         # ./lib/neo4j/active_node/persistence.rb:101:in `tap'
         # ./lib/neo4j/active_node/persistence.rb:101:in `create'
         # ./spec/e2e/label_spec.rb:87:in `block (5 levels) in <top (required)>'
         # ./spec/e2e/label_spec.rb:87:in `block (4 levels) in <top (required)>'
     # ./spec/e2e/label_spec.rb:87:in `block (4 levels) in <top (required)>'

Finished in 59.26 seconds
1015 examples, 3 failures, 4 pending
```

**Comment by cheerfulstoic** [comment]
Good catch, thanks!  I think perhaps this isn't just Neo4j 2.3.0 but an issue with non-jRuby version.  Check out Travis: https://travis-ci.org/neo4jrb/neo4j/builds/86952650

**Comment by cheerfulstoic** [comment]
Just getting back to this now, I think it is 2.3.0!  Tracking it down now, but this item in the [release notes](http://neo4j.com/release-notes/neo4j-2-3-0/) seems like it might be the issue:

"Status codes of the ‘transient’ classification now cause transactions to be rolled back. This resolves several outstanding issues where deadlocks, network errors and other similar problems required clients to manually roll back transactions. Only ClientError and ClientNotification classified status codes leave open transactions running."

**Comment by cheerfulstoic** [comment]
Ok, I think I've fixed this out.  I'd appreciate @subvertallchris taking a peek at https://github.com/neo4jrb/neo4j-core/commit/284ad1b1dd62c4d431620beaa48af8dc7cfc9287 though

The explanation: I think that now, in 2.3.0, most exceptions ("transient" exceptions) from Neo4j will now automatically fail the transaction and so we no longer need to make a final call to close it.  The gem was trying to make that final HTTP request and Neo4j was all like "Transaction?  What are you talking about?  I don't know about that transaction"

So, yay for fewer requests!

**Comment by subvertallchris** [comment]
That's really cool! Looks good to me!

**Comment by cheerfulstoic** [comment]
Just released 5.1.8 and 6.0.0.alpha.3 to fix.  Please re-open if that doesn't help
